Source/Core/Core/ScriptEngine.cpp

```diff
@@ -0,0 +1,945 @@
+  lua_setfield(L, -2, name);
+}
+
+static void init(lua_State* L)
+{
+  luaL_openlib(L, "dolphin", Lua::dolphinlib, 0);  // TODO Don't use openlib
+  int frame = lua_gettop(L);
+  lua_getglobal(L, "dolphin");
+
+  // ppc object
+  lua_newtable(L);
+  register_getset_object(L, "gpr", gpr_get, gpr_set);
+  register_getset_object(L, "sr", sr_get, sr_set);
+  register_getset_object(L, "spr", spr_get, spr_set);
+  register_ps_registers(L, "ps_u32", get_ps_u32, set_ps_u32);
+  register_ps_registers(L, "ps_f64", get_ps_f64, set_ps_f64);
+  new_getset_metatable(L, ppc_get, ppc_set);
+  lua_setmetatable(L, -2);
+
+  // dolphin["ppc"] = ppc
+  lua_setfield(L, -2, "ppc");
+
+  // memory access
+  register_getset_object(L, "mem_u8", read_u8, write_u8);
+  register_getset_object(L, "mem_u16", read_u16, write_u16);
+  register_getset_object(L, "mem_u32", read_u32, write_u32);
+  register_getset_object(L, "mem_f32", read_f32, write_f32);
+  register_getset_object(L, "mem_f64", read_f64, write_f64);
+
+  lua_pop(L, 1);
+  assert(lua_gettop(L) == frame);
+  luaL_openlibs(L);
+}
+
+}  // namespace Lua
+
+// Constructs a script from a target.
+Script::Script(std::string file_path) : m_file_path(std::move(file_path))
+{
+  Script::Load();
+}
+
+// Fully unloads a script destroying the context object, references to the script, and the Lua
+// state.
+Script::~Script()
+{
+  Unload();
+}
+
+Script::Script(Script&& other) noexcept : m_ctx(other.m_ctx)
+{
+  other.m_ctx = nullptr;
+}
+
+void Script::Load()
+{
+  if (m_ctx != nullptr)
+    return;
+  // Create Lua state.
+  lua_State* L = luaL_newstate();
+  Lua::init(L);
+  m_ctx = new Context(L);
+  fflush(stdout);
+  m_ctx->self = m_ctx;  // TODO This looks weird.
+  // Write the context reference to Lua.
+  lua_pushstring(L, Lua::s_context_registry_key);
+  lua_pushlightuserdata(L, reinterpret_cast<void*>(m_ctx));
+  lua_settable(L, LUA_REGISTRYINDEX);
+  // Load script.
+  int err = luaL_dofile(L, m_file_path.c_str());
+  if (err != 0)
+  {
+    const char* error_str = lua_tostring(L, -1);
+    PanicAlert("Error running Lua script: %s", error_str);
+    Unload();
+    return;
+  }
+}
+
+void Script::Unload()
+{
+  fflush(stdout);
+  delete m_ctx;
+  m_ctx = nullptr;
+}
+
+Script::Context::~Context()
+{
+  auto& hooks = PowerPC::script_hooks.m_hooks;
+  for (auto it = hooks.cbegin(); it != hooks.cend();)
+    if (it->second.ctx() == self)
+      hooks.erase(it++);
+    else
+      ++it;
+  lua_close(L);
+}
+
+void Script::Context::ExecuteHook(int lua_ref)
+{
+  int frame = lua_gettop(L);
+  lua_rawgeti(L, LUA_REGISTRYINDEX, lua_ref);
+  int err = lua_pcall(L, 0, 0, 0);
+  if (err != 0)
+  {
+    const char* error_str = lua_tostring(L, -1);
+    PanicAlert("Error running Lua hook: %s", error_str);
+  }
+  lua_settop(L, frame);
+}
+
+static std::vector<Script> s_scripts;
+
+void LoadScriptSection(const std::string& section, std::vector<ScriptTarget>& scripts,
+                       IniFile& localIni)
```

Comment:
```suggestion
                       IniFile& local_ini)
```

We use `snake_case` for variables and this file contains quite a bit that don't follow this.

Reply:
Renamed all `camelCase` variables I could find in ScriptEngine and ScriptWidget to `snake_case`.